pyo3-macros-backend/src/pymethod.rs

```diff
@@ -432,26 +432,38 @@ fn impl_traverse_slot(
     ctx: &Ctx,
 ) -> syn::Result<MethodAndSlotDef> {
     let Ctx { pyo3_path } = ctx;
-    if let (Some(py_arg), _) = split_off_python_arg(&spec.signature.arguments) {
+    let (py_arg, args) = split_off_python_arg(&spec.signature.arguments);
+    if let Some(py_arg) = py_arg {
         return Err(syn::Error::new_spanned(py_arg.ty, "__traverse__ may not take `Python`. \
-            Usually, an implementation of `__traverse__` should do nothing but calls to `visit.call`. \
-            Most importantly, safe access to the GIL is prohibited inside implementations of `__traverse__`, \
-            i.e. `Python::with_gil` will panic."));
+            Usually, an implementation of `__traverse__(&self, visit: PyVisit<'_>) -> Result<(), PyTraverseError>` \
+            should do nothing but calls to `visit.call`. Most importantly, safe access to the GIL is prohibited \
+            inside implementations of `__traverse__`, i.e. `Python::with_gil` will panic."));
     }
 
+    // check that the receiver does not try to smuggle an (implicit) `Python` token into here
     if let FnType::Fn(SelfType::TryFromBoundRef(span))
     | FnType::Fn(SelfType::Receiver {
         mutable: true,
         span,
     }) = spec.tp
     {
         bail_spanned! { span =>
-            "__traverse__ may not take a receiver other than `&self`. Usually, an implementation of `__traverse__` \
+            "__traverse__ may not take a receiver other than `&self`. Usually, an implementation of \
+            `__traverse__(&self, visit: PyVisit<'_>) -> Result<(), PyTraverseError>` \
             should do nothing but calls to `visit.call`. Most importantly, safe access to the GIL is prohibited \
             inside implementations of `__traverse__`, i.e. `Python::with_gil` will panic."
         }
     }
 
+    // check that we have a single argument (other than `&self`) of type `PyVisit`
+    if !matches!(args, [FnArg { ty: syn::Type::Path(p), .. }] if p.path.segments.last().map(|seg| seg.ident == "PyVisit").unwrap_or(false))
```

Comment:
Uff, while it would be nice to introduce error handling like this, I am not keen on match types by name, e.g. what if user code as 

```rust
use pyo3::pyclass::PyVisit as TraverseVisitor;
```

or

```rust
type TraverseVisitor = pyo3::pyclass::PyVisit;
```

We would fail valid code because we got confused by the superficial name change.

Reply:
Yeah, that is true. My thought was that it more likely to get the signature wrong, than someone renaming the type, but I agree that failing valid code is not great. I'll revert that bit

Reply:
Certainly more likely, but the failure mode of rejecting valid code is worse compared to bad error messages IMHO. Ideally we could have both but proc macros can only do so much.